File: ffi/src/share_token.rs
use crate::{error::Error, state::State};
use ouisync_lib::{network, ShareToken};

/// Returns the access mode of the given share token.
pub(crate) fn mode(token: ShareToken) -> u8 {
    token.access_mode().into()
}

/// Returns the info-hash of the repository corresponding to the share token formatted as hex
/// string.
pub(crate) fn info_hash(token: ShareToken) -> String {
    hex::encode(network::repository_info_hash(token.id()).as_ref())
}

pub(crate) fn suggested_name(token: ShareToken) -> String {
    token.suggested_name().to_owned()
}

/// Check if the repository is mirrored on the given server.
pub(crate) async fn mirror_exists(
    state: &State,
    token: ShareToken,
    host: &str,
) -> Result<bool, Error> {
    let config = state.get_remote_client_config().await?;
    Ok(ouisync_bridge::repository::mirror_exists(token.id(), config, host).await?)
}

File: bindings/swift/Ouisync/Sources/ShareToken.swift
import MessagePack


public extension Client {
    func shareToken(fromString value: String) async throws -> ShareToken {
        try await ShareToken(self, invoke("share_token_normalize", with: .string(value)))
    }
}


public struct ShareToken: Secret {
    // FIXME: should this retain the Client unless cast to string?
    let client: Client
    /// The (nominally secret) token value that should be stored or shared somewhere safe
    public let string: String
    public var value: MessagePackValue { .string(string) }

    init(_ client: Client, _ value: MessagePackValue) throws {
        self.client = client
        string = try value.stringValue.orThrow
    }
}


public extension ShareToken {
    /// The repository name suggested from this token.
    var suggestedName: String { get async throws {
        try await client.invoke("share_token_get_suggested_name", with: value).stringValue.orThrow
    } }

    var infoHash: String { get async throws {
        try await client.invoke("share_token_get_info_hash", with: value).stringValue.orThrow
    } }

    /// The access mode this token provides.
    var accessMode: AccessMode { get async throws {
        try await AccessMode(rawValue: client.invoke("share_token_get_access_mode",
                                                     with: value).uint8Value.orThrow).orThrow
    } }

    /// Check if the repository of this share token is mirrored on the cache server.
    func mirrorExists(at host: String) async throws -> Bool {
        try await client.invoke("share_token_mirror_exists",
                                with: ["share_token": value,
                                       "host": .string(host)]).boolValue.orThrow
    }
}
